{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "03faeb86",
   "metadata": {},
   "source": [
    "## AMR in genomes from mysterious city\n",
    "First lets upload the mysterious city AMR presence-absence table\n",
    "We have 143 genomes (rows) with 505 AMR markers detected"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d38dccee",
   "metadata": {},
   "source": [
    "We have many tables of CARD-manual. Since mysterious samples just register presence and absence we divide our tables in Count tables and Presence tables.\n",
    "\n",
    "The list of such tables are in: https://github.com/FranciscoSNR/camda2023/tree/main/Taxonomy\n",
    "This table were made by Anthon."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c09ed271",
   "metadata": {},
   "source": [
    "Todas las tablas poseen las siguientes columnas:\n",
    "Genus . Género predicho por RGI (el anotador de genes de resistencia utilizado).\n",
    "Gene . Nombre del gen de resistencia.\n",
    "ARO . Identificador único del gen en la base de datos de CARD.\n",
    "El resto de las columnas corresponden a los ensamblados analizados.\n",
    "Las tablas están clasificadas en cuatro categorías:\n",
    "\n",
    "SampleAssemblies . Genes de resistencia en ensamblados por muestra. El formato de las\n",
    "columnas es AABBBCC , done AA corresponde al año, BBB es el nombre de la ciudad con tres letras, y\n",
    "CC es el número de muestra.\n",
    "\n",
    "Coassemblies . Genes de resistencia en coensamblados por ciudad. El formato de las columnas es\n",
    "BBB (es decir, el nombre de la ciudad con tres letras).\n",
    "\n",
    "CoassembliesExtracted . Genes de resistencia en los contigs extraídos de los coensamblados\n",
    "por ciudad que Kraken2 clasificó como Enterobacter, Escherichia o Klebsiella. El formato de las\n",
    "columnas es BBB_Dd , donde BBB es el nombre de tres letras de la ciudad y Dd es el nombre de la\n",
    "especie ( En : Enterobacter, Es : Escherichia, Kl : Klebsiella)\n",
    "\n",
    "ExtractedReadAssemblies . Genes de resistencia en los ensamblados de las lecturas por ciudad\n",
    "que Kraken2 clasificó como Enterobacter, Escherichia o Klebsiella. El formato del nombre de las\n",
    "columnas es idéntico al de CoassembliesExtracted .\n",
    "\n",
    "Nota. Las tablas de Coassemblies , CoassembliesExtracted y ExtractedReadAssemblies\n",
    "sólo incluyen ciudades de Estados Unidos.\n",
    "Las tablas de SampleAssemblies incluyen a todas las ciudades.\n",
    "\n",
    "Cada una de estas categorías posee dos tablas marcadas como Strict y dos como Full :\n",
    "\n",
    "Las tablas Strict contienen genes de resistencia que RGI identificó como confiables.\n",
    "\n",
    "Las tablas Full incluyen a todos los genes de resistencia, incluyendo aquellos cuyo nivel de\n",
    "confianza es menor.\n",
    "\n",
    "Las tablas también se encuentran clasificadas en Counts y en Presence :\n",
    "Las tablas Counts guardan los conteos de cada gen por ensamble analizado.\n",
    "Las tablas Presence almacenan las presencias (1s) y ausencias (0s) por ensamble analizado. Se\n",
    "contruyeron a partir de las tablas de Counts : se marcan como presentes aquéllos genes que están\n",
    "presentes por lo menos una vez."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "60205d7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   ID  ARO:3000167  ARO:3002847  mrkb  enta  fepa  ARO:3004122  cyca  csge  \\\n",
      "0  S1            0            0     0     0     0            0     1     0   \n",
      "1  S2            0            0     0     0     0            0     1     0   \n",
      "2  S3            0            0     0     0     0            0     1     0   \n",
      "3  S4            0            0     0     1     0            0     1     0   \n",
      "4  S5            0            0     0     0     0            0     1     0   \n",
      "\n",
      "   iucc  ...  teml100  scintssj  wcst  sfax  kpsd  mdlb  ARO:3001084  tssj  \\\n",
      "0     0  ...        0         0     0     0     0     0            0     0   \n",
      "1     0  ...        0         0     0     0     0     0            0     0   \n",
      "2     0  ...        0         0     0     0     0     0            0     0   \n",
      "3     0  ...        0         0     0     0     0     0            0     0   \n",
      "4     0  ...        0         0     0     0     0     0            0     0   \n",
      "\n",
      "   icmftssm  acrr  \n",
      "0         0     0  \n",
      "1         0     0  \n",
      "2         0     0  \n",
      "3         0     0  \n",
      "4         0     0  \n",
      "\n",
      "[5 rows x 506 columns]\n",
      "Number of rows: 146\n",
      "Number of columns: 506\n",
      "0       74\n",
      "1       79\n",
      "2       85\n",
      "3       85\n",
      "4       76\n",
      "      ... \n",
      "141    125\n",
      "142    126\n",
      "143    126\n",
      "144    118\n",
      "145    115\n",
      "Length: 146, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "## 505 combined CARD-manual Ids from mysterious samples   ### This table is incorrect, more ones are required by roow\n",
    "url = 'https://raw.githubusercontent.com/ccm-bioinfo/cambda2023/main/06_amr_resistance/data/amr_mistery_table20230622_aroIds.tsv'\n",
    "# Read the TSV file into a DataFrame\n",
    "df_CM_mysterious = pd.read_csv(url, delimiter='\\t')\n",
    "# Rename the first column\n",
    "df_CM_mysterious.columns.values[0] = 'ID'\n",
    "\n",
    "# Print the DataFrame\n",
    "print(df_CM_mysterious.head())\n",
    "\n",
    "# Get the dimensions of the DataFrame\n",
    "rows_df_CM_mysterious, columns_df_CM_mysterious = df_CM_mysterious.shape\n",
    "\n",
    "# Print the dimensions\n",
    "print(\"Number of rows:\", rows_df_CM_mysterious)\n",
    "print(\"Number of columns:\", columns_df_CM_mysterious)\n",
    "# Count non-zero values by row\n",
    "non_zero_counts = (df_CM_mysterious != 0).sum(axis=1)\n",
    "\n",
    "# Print the result\n",
    "print(non_zero_counts)\n",
    "#print (df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ebd71cd",
   "metadata": {},
   "source": [
    "The same datafram with spices. Then add a column called city"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "96e97237",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   ID                  Species        City AST-based group  ARO:3000167  \\\n",
      "0  S1  Enterobacter hormaechei  mysterious   3GC-resistant            0   \n",
      "1  S2  Enterobacter hormaechei  mysterious          CP CRE            0   \n",
      "2  S3  Enterobacter hormaechei  mysterious          CP CRE            0   \n",
      "3  S4  Enterobacter hormaechei  mysterious          CP CRE            0   \n",
      "4  S5  Enterobacter hormaechei  mysterious          CP CRE            0   \n",
      "\n",
      "   ARO:3002847  mrkb  enta  fepa  ARO:3004122  ...  teml100  scintssj  wcst  \\\n",
      "0            0     0     0     0            0  ...        0         0     0   \n",
      "1            0     0     0     0            0  ...        0         0     0   \n",
      "2            0     0     0     0            0  ...        0         0     0   \n",
      "3            0     0     1     0            0  ...        0         0     0   \n",
      "4            0     0     0     0            0  ...        0         0     0   \n",
      "\n",
      "   sfax  kpsd  mdlb  ARO:3001084  tssj  icmftssm  acrr  \n",
      "0     0     0     0            0     0         0     0  \n",
      "1     0     0     0            0     0         0     0  \n",
      "2     0     0     0            0     0         0     0  \n",
      "3     0     0     0            0     0         0     0  \n",
      "4     0     0     0            0     0         0     0  \n",
      "\n",
      "[5 rows x 509 columns]\n"
     ]
    }
   ],
   "source": [
    "## Adding metadata from such as species to mysterious samples\n",
    "url = 'https://raw.githubusercontent.com/ccm-bioinfo/cambda2023/main/06_amr_resistance/codigos/data_preparation/data/amr_patterns.tsv'  # Replace with the actual GitHub URL of the CSV file\n",
    "\n",
    "# Read the TSV file into a DataFrame\n",
    "df_metadata_mysterious = pd.read_csv(url, delimiter='\\t')\n",
    "new_column = df_metadata_mysterious['ID']\n",
    "df_metadata_mysterious.insert(2, 'City', new_column)\n",
    "df_metadata_mysterious['City'] = 'mysterious'\n",
    "df_metadata_mysterious=df_metadata_mysterious.iloc[:, :4]\n",
    "\n",
    "# Print the DataFrame\n",
    "#print(df_metadata_mysterious.head())\n",
    "# Perform the join based on the 'Id' column\n",
    "mysterious_df = pd.merge( df_metadata_mysterious,df_CM_mysterious, on='ID')\n",
    "\n",
    "\n",
    "# Print the merged DataFrame\n",
    "print(mysterious_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caa2cecb",
   "metadata": {},
   "source": [
    "## AMR in metagenomes  \n",
    "### Extracted by taxonomy - Binned by city - Assembled - AMR profiled\n",
    "\n",
    "As we mantion before, there are many AMR-CARD tables that were gotten with diferent methods.\n",
    "Let explore some of them.\n",
    "\n",
    "First we focus on precense and absent tables because mysterius sample is presented in that way. \n",
    "\n",
    "\n",
    "%We found 492 CARD Ids ( columns) not all of them are in the mysterious city samples.  \n",
    "%The 12 rows are the genus _Escherichia_, _Enterobacter_ and _Klebsiella_ in the 6 US cities were AMR CARD models were found.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "298ef49a",
   "metadata": {},
   "source": [
    "We start with Coassemblies_Strict_Presence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3279db15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARO     ARO:3001821   ARO:3001842   ARO:3006267   ARO:3006268   ARO:3001831  \\\n",
      "Genus  Enterobacter  Enterobacter  Enterobacter  Enterobacter  Enterobacter   \n",
      "Gene          ACT-1        ACT-21        ACT-73        ACT-74         ACT-8   \n",
      "BAL               0             0             0             0             0   \n",
      "DEN               0             0             0             0             0   \n",
      "MIN               0             0             0             0             0   \n",
      "\n",
      "ARO     ARO:3000518                ARO:3004042  \\\n",
      "Genus  Enterobacter               Enterobacter   \n",
      "Gene            CRP  Enterobacter cloacae acrA   \n",
      "BAL               0                          0   \n",
      "DEN               0                          0   \n",
      "MIN               0                          0   \n",
      "\n",
      "ARO                                          ARO:3003369  \\\n",
      "Genus                                       Enterobacter   \n",
      "Gene   Escherichia coli EF-Tu mutants conferring resi...   \n",
      "BAL                                                    0   \n",
      "DEN                                                    0   \n",
      "MIN                                                    0   \n",
      "\n",
      "ARO                                          ARO:3003890  \\\n",
      "Genus                                       Enterobacter   \n",
      "Gene   Escherichia coli UhpT with mutation conferring...   \n",
      "BAL                                                    0   \n",
      "DEN                                                    0   \n",
      "MIN                                                    0   \n",
      "\n",
      "ARO                                          ARO:3003378  ... ARO:3000516  \\\n",
      "Genus                                       Enterobacter  ...  Klebsiella   \n",
      "Gene   Escherichia coli marR mutant conferring antibi...  ...        emrR   \n",
      "BAL                                                    0  ...           0   \n",
      "DEN                                                    0  ...           1   \n",
      "MIN                                                    0  ...           0   \n",
      "\n",
      "ARO   ARO:3005047 ARO:3003209 ARO:3000263 ARO:3000793 ARO:3000794 ARO:3001216  \\\n",
      "Genus  Klebsiella  Klebsiella  Klebsiella  Klebsiella  Klebsiella  Klebsiella   \n",
      "Gene         eptB       fosA5        marA        mdtB        mdtC        mdtH   \n",
      "BAL             0           0           0           0           1           0   \n",
      "DEN             1           1           1           1           0           0   \n",
      "MIN             0           0           0           0           0           0   \n",
      "\n",
      "ARO   ARO:3003922 ARO:3003923 ARO:3007014  \n",
      "Genus  Klebsiella  Klebsiella  Klebsiella  \n",
      "Gene         oqxA        oqxB        qacJ  \n",
      "BAL             0           0           0  \n",
      "DEN             0           1           0  \n",
      "MIN             0           0           0  \n",
      "\n",
      "[5 rows x 97 columns]\n"
     ]
    }
   ],
   "source": [
    "# CARD table for Extracted by Taxonomy , binned by city\n",
    "url = 'https://raw.githubusercontent.com/FranciscoSNR/camda2023/main/Taxonomy/Coassemblies/Coassemblies_Strict_Presence.tsv' \n",
    "# Read the TSV file into a DataFrame\n",
    "\n",
    "\n",
    "df_ET_BC_AMR_CARD = pd.read_csv(url, delimiter='\\t').transpose()\n",
    "\n",
    "# Get the values from the third row\n",
    "third_row = df_ET_BC_AMR_CARD.iloc[2]\n",
    "#print(third_row)\n",
    "df_ET_BC_AMR_CARD.columns=third_row\n",
    "\n",
    "# Add a prefix to column names\n",
    "prefix = 'ARO:'  # Replace with the desired prefix\n",
    "df_ET_BC_AMR_CARD = df_ET_BC_AMR_CARD.add_prefix(prefix)\n",
    "\n",
    "### Las part was no required this time.\n",
    "# Remove a specific row by index\n",
    "index_to_remove = \"ARO\"  # Replace with the index of the row you want to remove\n",
    "df_ET_BC_AMR_CARD = df_ET_BC_AMR_CARD.drop(index=index_to_remove)\n",
    "\n",
    "print(df_ET_BC_AMR_CARD.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69c947d8",
   "metadata": {},
   "source": [
    "Now we will get the CARD Ids intersection with IDs in mysterious city samples, only 57 were found.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "42a05a2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARO     db   ARO:3000516 ARO:3000516  ARO:3000792  ARO:3003550  \\\n",
      "0    Genus  Enterobacter  Klebsiella  Escherichia  Escherichia   \n",
      "1     Gene          emrR        emrR         mdtA         mdtP   \n",
      "2      BAL             1           0            0            0   \n",
      "3      DEN             0           1            0            0   \n",
      "4      MIN             0           0            0            0   \n",
      "\n",
      "ARO                 ARO:3004588  ARO:3000027   ARO:3000823  ARO:3000833  \\\n",
      "0                    Klebsiella  Escherichia  Enterobacter  Escherichia   \n",
      "1    Klebsiella pneumoniae KpnG         emrA          ramA         evgS   \n",
      "2                             0            0             0            0   \n",
      "3                             1            0             0            0   \n",
      "4                             0            0             0            0   \n",
      "\n",
      "ARO   ARO:3000676  ...  ARO:3001216 ARO:3001216   ARO:3000828 ARO:3000828  \\\n",
      "0    Enterobacter  ...  Escherichia  Klebsiella  Enterobacter  Klebsiella   \n",
      "1            H-NS  ...         mdtH        mdtH          baeR        baeR   \n",
      "2               0  ...            0           0             0           1   \n",
      "3               0  ...            0           0             0           0   \n",
      "4               0  ...            0           0             0           0   \n",
      "\n",
      "ARO   ARO:3000830  ARO:3000830 ARO:3000830  ARO:3000254  \\\n",
      "0    Enterobacter  Escherichia  Klebsiella  Escherichia   \n",
      "1            cpxA         cpxA        cpxA         emrY   \n",
      "2               0            1           0            0   \n",
      "3               0            0           1            0   \n",
      "4               0            0           0            0   \n",
      "\n",
      "ARO                   ARO:3004122   ARO:3002804  \n",
      "0                      Klebsiella  Enterobacter  \n",
      "1    Klebsiella pneumoniae OmpK37         FosA2  \n",
      "2                               0             0  \n",
      "3                               0             0  \n",
      "4                               0             0  \n",
      "\n",
      "[5 rows x 59 columns]\n"
     ]
    }
   ],
   "source": [
    "# Create an empty column with the desired name\n",
    "new_column = df_ET_BC_AMR_CARD.index\n",
    "\n",
    "# Get the intersection of column names\n",
    "common_columns = list(set(df_CM_mysterious.columns).intersection(set(df_ET_BC_AMR_CARD.columns)))\n",
    "\n",
    "# Keep only the common columns in both DataFrames\n",
    "#df1 = df_CM_mysterious[common_columns]\n",
    "df_ET_BC_AMR_CARD = df_ET_BC_AMR_CARD[common_columns]\n",
    "\n",
    "# Print the updated DataFrames with only the common columns\n",
    "# Assign the empty column as the first column in the DataFrame\n",
    "df_ET_BC_AMR_CARD.insert(0, 'db', new_column)\n",
    "df_ET_BC_AMR_CARD=df_ET_BC_AMR_CARD.reset_index(drop=True)\n",
    "\n",
    "print(df_ET_BC_AMR_CARD.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4748e7d7",
   "metadata": {},
   "source": [
    "Now we want to load the blast table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2da5ba25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       db  tssa  kpsm  gspm  ykgkecpr  wcag  hcp2tssd2  kpse  \\\n",
      "0  BAL_En     0     0     0         0     0          0     0   \n",
      "1  BAL_Es     0     0     0         0     0          0     0   \n",
      "2  BAL_Kl     0     0     0         0     0          0     0   \n",
      "3  DEN_En     0     0     0         0     0          0     0   \n",
      "4  DEN_Es     0     0     0         0     0          0     0   \n",
      "\n",
      "   aac630_aac6ibfusionprotein  hlyd  ...  chuy  flii  papk  papj  mrkh  flia  \\\n",
      "0                           0     0  ...     0     1     0     0     0     0   \n",
      "1                           0     0  ...     0     0     0     0     0     0   \n",
      "2                           0     0  ...     0     0     0     0     0     0   \n",
      "3                           0     0  ...     0     0     0     0     0     0   \n",
      "4                           0     0  ...     0     0     0     0     0     0   \n",
      "\n",
      "   ybts  clbo  ibea  papc  \n",
      "0     0     0     0     0  \n",
      "1     0     0     0     0  \n",
      "2     0     0     0     0  \n",
      "3     0     0     0     0  \n",
      "4     0     0     0     0  \n",
      "\n",
      "[5 rows x 326 columns]\n"
     ]
    }
   ],
   "source": [
    "url2='https://raw.githubusercontent.com/ccm-bioinfo/cambda2023/main/06_amr_resistance/data/amr_counts_complement_b3_20230629.tsv'\n",
    "# this link is the result given by blast our 325 ids database vs cambda metagenomic samples\n",
    "\n",
    "\n",
    "# Read the TSV file into a DataFrame\n",
    "df_ET_BC_AMR_BLASTn = pd.read_csv(url2, delimiter='\\t')\n",
    "df_ET_BC_AMR_BLASTn['db'] = df_ET_BC_AMR_BLASTn['db'].str.split('.').str[0]\n",
    "\n",
    "# Create an empty column with the desired name\n",
    "new_column = df_ET_BC_AMR_BLASTn['db']\n",
    "\n",
    "# Print the DataFrame\n",
    "#print(df_ET_BC_AMR_BLASTn)\n",
    "\n",
    "# Get the intersection of column names\n",
    "common_columns = list(set(df_CM_mysterious.columns).intersection(df_ET_BC_AMR_BLASTn))\n",
    "\n",
    "# Keep only the common columns in both DataFrames\n",
    "#df1 = df_CM_mysterious[common_columns]\n",
    "df_ET_BC_AMR_BLASTn= df_ET_BC_AMR_BLASTn[common_columns]\n",
    "# Transform DataFrame to boolean\n",
    "df_boolean = df_ET_BC_AMR_BLASTn.astype(bool)\n",
    "# Convert boolean values back to numeric\n",
    "df_ET_BC_AMR_BLASTn = df_boolean.astype(int)\n",
    "\n",
    "\n",
    "# Assign the empty column as the first column in the DataFrame\n",
    "df_ET_BC_AMR_BLASTn.insert(0, 'db', new_column)\n",
    "df_ET_BC_AMR_BLASTn=df_ET_BC_AMR_BLASTn.reset_index(drop=True)\n",
    "\n",
    "# Print the updated DataFrames with only the common columns\n",
    "#print(df1)\n",
    "print(df_ET_BC_AMR_BLASTn.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722bc43a",
   "metadata": {},
   "source": [
    "We want to joint the AMR_BLASTn table and AMR_CARD table horizontaly. However the structure of our tables are diferent."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d1c6c95",
   "metadata": {},
   "source": [
    "We will transform the AMR_CARD table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9ae20d0",
   "metadata": {},
   "source": [
    "Our problem is that the have diferenmts formats. We want secondone looks like firstone.\n",
    "\n",
    "In particular :\n",
    "\n",
    "1) the first column has the name \"db\" and it has 18 rows of the form city_spice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "38ec230a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data=[]\n",
    "new_row_En=pd.DataFrame(data)\n",
    "for c in range(2,8):\n",
    "    city=df_ET_BC_AMR_CARD[\"db\"]\n",
    "    new_row_En[\"db\"]=city[c]+\"_\"+\"En\"\n",
    "    data.append(city[c]+\"_\"+\"En\")\n",
    "    new_row_En[\"db\"]=city[c]+\"_\"+\"Es\"\n",
    "    data.append(city[c]+\"_\"+\"Es\")\n",
    "    new_row_En[\"db\"]=city[c]+\"_\"+\"Kl\"\n",
    "    data.append(city[c]+\"_\"+\"Kl\")\n",
    "df_ET_BC_AMR_CARD_mod=pd.DataFrame(data)\n",
    "df_ET_BC_AMR_CARD_mod=df_ET_BC_AMR_CARD_mod.rename(columns={0: 'db'})\n",
    "#print(df_ET_BC_AMR_CARD_mod)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11f41e28",
   "metadata": {},
   "source": [
    "Got it\n",
    "\n",
    "Now We need to spllit range of 18 in 3 sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c07646bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 3, 6, 9, 12, 15] [1, 4, 7, 10, 13, 16] [2, 5, 8, 11, 14, 17]\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Initialize a defaultdict to group numbers based on their remainder when divided by 3\n",
    "grouped_nums = defaultdict(list)\n",
    "\n",
    "# Divide range(18) into 3 sets based on remainder when divided by 3\n",
    "for num in range(18):\n",
    "    remainder = num % 3\n",
    "    grouped_nums[remainder].append(num)\n",
    "List_num=list(grouped_nums.items())\n",
    "List_0=List_num[0][1]\n",
    "List_1=List_num[1][1]\n",
    "List_2=List_num[2][1]\n",
    "print(List_0,List_1,List_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fca1448",
   "metadata": {},
   "source": [
    "2) We want a column for each gene in the 3th roe in the table modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e743612c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       db  ARO:3000516_1  ARO:3000516_2  ARO:3000792_3  ARO:3003550_4  \\\n",
      "0  BAL_En              1              0              0              0   \n",
      "1  BAL_Es              0              0              0              0   \n",
      "2  BAL_Kl              0              0              0              0   \n",
      "3  DEN_En              0              0              0              0   \n",
      "4  DEN_Es              0              0              0              0   \n",
      "\n",
      "   ARO:3004588_5  ARO:3000027_6  ARO:3000823_7  ARO:3000833_8  ARO:3000676_9  \\\n",
      "0              0              0              0              0              0   \n",
      "1              0              0              0              0              0   \n",
      "2              0              0              0              0              0   \n",
      "3              0              0              0              0              0   \n",
      "4              0              0              0              0              0   \n",
      "\n",
      "   ...  ARO:3001216_49  ARO:3001216_50  ARO:3000828_51  ARO:3000828_52  \\\n",
      "0  ...               0               0               0               0   \n",
      "1  ...               0               0               0               0   \n",
      "2  ...               0               0               0               1   \n",
      "3  ...               0               0               0               0   \n",
      "4  ...               0               0               0               0   \n",
      "\n",
      "   ARO:3000830_53  ARO:3000830_54  ARO:3000830_55  ARO:3000254_56  \\\n",
      "0               0               0               0               0   \n",
      "1               0               1               0               0   \n",
      "2               0               0               0               0   \n",
      "3               0               0               0               0   \n",
      "4               0               0               0               0   \n",
      "\n",
      "   ARO:3004122_57  ARO:3002804_58  \n",
      "0               0               0  \n",
      "1               0               0  \n",
      "2               0               0  \n",
      "3               0               0  \n",
      "4               0               0  \n",
      "\n",
      "[5 rows x 59 columns]\n"
     ]
    }
   ],
   "source": [
    "names_of_col=[]\n",
    "new_col_zeros=[]\n",
    "for i in range(18):\n",
    "    new_col_zeros.append(0)\n",
    "    \n",
    "for i in range(1,len(df_ET_BC_AMR_CARD.columns)):\n",
    "    new_col=[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "    #new_col=new_col_ceros\n",
    "    #print(\"new_col_ceros  entering for\",new_col_ceros)\n",
    "    #print(\"new_col  entering for\",new_col)\n",
    "\n",
    "    if df_ET_BC_AMR_CARD.iloc[0,i]==\"Enterobacter\":\n",
    "        modulo=0\n",
    "    elif df_ET_BC_AMR_CARD.iloc[0,i]==\"Escherichia\":\n",
    "        modulo=1\n",
    "    elif df_ET_BC_AMR_CARD.iloc[0,i]==\"Klebsiella\":\n",
    "        modulo=2\n",
    "        \n",
    "    for j in range(6):\n",
    "        if modulo==0:\n",
    "            if  df_ET_BC_AMR_CARD.iloc[j+2,i]==1:\n",
    "                new_col[List_0[j]]=1\n",
    "            else:\n",
    "                new_col[List_0[j]]=0\n",
    "        elif modulo==1:\n",
    "            if  df_ET_BC_AMR_CARD.iloc[j+2,i]==1:\n",
    "                new_col[List_1[j]]=1\n",
    "            else:\n",
    "                new_col[List_0[j]]=0\n",
    "        elif modulo==2:\n",
    "            if  df_ET_BC_AMR_CARD.iloc[j+2,i]==1:\n",
    "                new_col[List_2[j]]=1\n",
    "            else:\n",
    "                new_col[List_0[j]]=0\n",
    "    #print(\"new_col\", new_col)                \n",
    "    df_ET_BC_AMR_CARD_mod[df_ET_BC_AMR_CARD.columns[i]+\"_\"+str(i)]=new_col\n",
    "        \n",
    "        #restart the column with zeros\n",
    "    #print(\"new_col_ceros\",new_col_ceros)\n",
    "\n",
    "    new_col=new_col_zeros\n",
    "    #print(\"ne_col after equal to new_col_ceros\",new_col)\n",
    "    #new_col.append(df_ET_BC_AMR_CARD.iloc[2,i]+\"_\"+i)\n",
    "    #if df_ET_BC_AMR_CARD.iloc[2,i] not in names_of_col:\n",
    "    #    names_of_col.append(df_ET_BC_AMR_CARD.iloc[2,i])\n",
    "print(df_ET_BC_AMR_CARD_mod.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5618d96",
   "metadata": {},
   "source": [
    "we will to \"colaps\" or sum the columns by gene. if 2 columns start in the same way (x), sum them and get a new col called x delait the columns of the form \"x_number\" and at \"x\" to the end.\n",
    "\n",
    "get the name of the genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5a725adc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       db  ARO:3000516  ARO:3000792  ARO:3003550  ARO:3004588  ARO:3000027  \\\n",
      "0  BAL_En            1            0            0            0            0   \n",
      "1  BAL_Es            0            0            0            0            0   \n",
      "2  BAL_Kl            0            0            0            0            0   \n",
      "3  DEN_En            0            0            0            0            0   \n",
      "4  DEN_Es            0            0            0            0            0   \n",
      "\n",
      "   ARO:3000823  ARO:3000833  ARO:3000676  ARO:3005044  ...  ARO:3005059  \\\n",
      "0            0            0            0            0  ...            0   \n",
      "1            0            0            0            0  ...            0   \n",
      "2            0            0            0            0  ...            0   \n",
      "3            0            0            0            0  ...            0   \n",
      "4            0            0            0            0  ...            0   \n",
      "\n",
      "   ARO:3003922  ARO:3003952  ARO:3000216  ARO:3001216  ARO:3000828  \\\n",
      "0            0            0            0            0            0   \n",
      "1            0            0            0            0            0   \n",
      "2            0            0            0            0            1   \n",
      "3            0            0            0            0            0   \n",
      "4            0            0            0            0            0   \n",
      "\n",
      "   ARO:3000830  ARO:3000254  ARO:3004122  ARO:3002804  \n",
      "0            0            0            0            0  \n",
      "1            1            0            0            0  \n",
      "2            0            0            0            0  \n",
      "3            0            0            0            0  \n",
      "4            0            0            0            0  \n",
      "\n",
      "[5 rows x 42 columns]\n"
     ]
    }
   ],
   "source": [
    "List_of_names=[]\n",
    "for name in df_ET_BC_AMR_CARD_mod.columns[1:]:\n",
    "    real_name=name.split(\"_\")[0]\n",
    "    if real_name in List_of_names:\n",
    "        True\n",
    "    else:\n",
    "        List_of_names.append(real_name)\n",
    "#print(List_of_names)\n",
    "\n",
    "\n",
    "#redo the first column as before\n",
    "data=[]\n",
    "new_row_En=pd.DataFrame(data)\n",
    "#CARD=pd.DataFrame(data)\n",
    "for c in range(2,8):\n",
    "    city=df_ET_BC_AMR_CARD[\"db\"]\n",
    "    new_row_En[\"db\"]=city[c]+\"_\"+\"En\"\n",
    "    data.append(city[c]+\"_\"+\"En\")\n",
    "    new_row_En[\"db\"]=city[c]+\"_\"+\"Es\"\n",
    "    data.append(city[c]+\"_\"+\"Es\")\n",
    "    new_row_En[\"db\"]=city[c]+\"_\"+\"Kl\"\n",
    "    data.append(city[c]+\"_\"+\"Kl\")\n",
    "df_ET_BC_AMR_CARD_mod_2=pd.DataFrame(data)\n",
    "df_ET_BC_AMR_CARD_mod_2=df_ET_BC_AMR_CARD_mod.rename(columns={0: 'db'})\n",
    "\n",
    "\n",
    "#collect the columns that start the same, add them, add them to the end\n",
    "for real_name in List_of_names:\n",
    "    for name in df_ET_BC_AMR_CARD_mod.columns[1:]:\n",
    "        if real_name==name.split(\"_\")[0]:\n",
    "            real_name_columns = [col for col in df_ET_BC_AMR_CARD_mod.columns[1:] if col.startswith(real_name)]\n",
    "            sum_of_real_name_columns=df_ET_BC_AMR_CARD_mod[real_name_columns].sum(axis=1)\n",
    "            df_ET_BC_AMR_CARD_mod_2[real_name]=sum_of_real_name_columns\n",
    "            \n",
    "#delate columns of thr form gen_number\n",
    "for name in df_ET_BC_AMR_CARD_mod.columns[1:]:\n",
    "    del df_ET_BC_AMR_CARD_mod_2[name]\n",
    "print(df_ET_BC_AMR_CARD_mod_2.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "796b5f40",
   "metadata": {},
   "source": [
    "We got it, bouth AMR-CARD and AMR-BLASTn tables have \"the same format\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77bdcf3f",
   "metadata": {},
   "source": [
    "Now lets horizontally combined the two datasets, so that for each city we will have 325+41 =366 AMR IDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "44cd13d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       ID  ARO:3000516  ARO:3000792  ARO:3003550  ARO:3004588  ARO:3000027  \\\n",
      "0  BAL_En            1            0            0            0            0   \n",
      "1  BAL_Es            0            0            0            0            0   \n",
      "2  BAL_Kl            0            0            0            0            0   \n",
      "3  DEN_En            0            0            0            0            0   \n",
      "4  DEN_Es            0            0            0            0            0   \n",
      "\n",
      "   ARO:3000823  ARO:3000833  ARO:3000676  ARO:3005044  ...  chuy  flii  papk  \\\n",
      "0            0            0            0            0  ...     0     1     0   \n",
      "1            0            0            0            0  ...     0     0     0   \n",
      "2            0            0            0            0  ...     0     0     0   \n",
      "3            0            0            0            0  ...     0     0     0   \n",
      "4            0            0            0            0  ...     0     0     0   \n",
      "\n",
      "   papj  mrkh  flia  ybts  clbo  ibea  papc  \n",
      "0     0     0     0     0     0     0     0  \n",
      "1     0     0     0     0     0     0     0  \n",
      "2     0     0     0     0     0     0     0  \n",
      "3     0     0     0     0     0     0     0  \n",
      "4     0     0     0     0     0     0     0  \n",
      "\n",
      "[5 rows x 367 columns]\n"
     ]
    }
   ],
   "source": [
    "merged_ETBCMet_df = pd.merge(df_ET_BC_AMR_CARD_mod_2, df_ET_BC_AMR_BLASTn, on='db')\n",
    "merged_ETBCMet_df.rename(columns={'db': 'ID'}, inplace=True)\n",
    "print(merged_ETBCMet_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aab88c2d",
   "metadata": {},
   "source": [
    "Now we joint vertically mysterious sample (146 rows) with oru AMR jointed table (18 rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6bfdf952",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   ID  ARO:3000167  ARO:3002847  mrkb  enta  fepa  ARO:3004122  cyca  csge  \\\n",
      "0  S1          0.0          0.0     0     0     0            0     1     0   \n",
      "1  S2          0.0          0.0     0     0     0            0     1     0   \n",
      "2  S3          0.0          0.0     0     0     0            0     1     0   \n",
      "3  S4          0.0          0.0     0     1     0            0     1     0   \n",
      "4  S5          0.0          0.0     0     0     0            0     1     0   \n",
      "\n",
      "   iucc  ...  teml100  scintssj  wcst  sfax  kpsd  mdlb  ARO:3001084  tssj  \\\n",
      "0     0  ...        0         0     0     0     0     0          0.0     0   \n",
      "1     0  ...        0         0     0     0     0     0          0.0     0   \n",
      "2     0  ...        0         0     0     0     0     0          0.0     0   \n",
      "3     0  ...        0         0     0     0     0     0          0.0     0   \n",
      "4     0  ...        0         0     0     0     0     0          0.0     0   \n",
      "\n",
      "   icmftssm  acrr  \n",
      "0         0     0  \n",
      "1         0     0  \n",
      "2         0     0  \n",
      "3         0     0  \n",
      "4         0     0  \n",
      "\n",
      "[5 rows x 506 columns]\n"
     ]
    }
   ],
   "source": [
    "concatenated_df = pd.concat([df_CM_mysterious, merged_ETBCMet_df], axis=0, ignore_index=True).fillna(0)\n",
    "print(concatenated_df.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4b7b0e7",
   "metadata": {},
   "source": [
    "However. We want to extend that table with metadata of the mysterious ssample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "da8dacc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   ID                  Species        City AST-based group  ARO:3000167  \\\n",
      "0  S1  Enterobacter hormaechei  mysterious   3GC-resistant          0.0   \n",
      "1  S2  Enterobacter hormaechei  mysterious          CP CRE          0.0   \n",
      "2  S3  Enterobacter hormaechei  mysterious          CP CRE          0.0   \n",
      "3  S4  Enterobacter hormaechei  mysterious          CP CRE          0.0   \n",
      "4  S5  Enterobacter hormaechei  mysterious          CP CRE          0.0   \n",
      "\n",
      "   ARO:3002847  mrkb  enta  fepa  ARO:3004122  ...  teml100  scintssj  wcst  \\\n",
      "0          0.0     0     0     0            0  ...        0         0     0   \n",
      "1          0.0     0     0     0            0  ...        0         0     0   \n",
      "2          0.0     0     0     0            0  ...        0         0     0   \n",
      "3          0.0     0     1     0            0  ...        0         0     0   \n",
      "4          0.0     0     0     0            0  ...        0         0     0   \n",
      "\n",
      "   sfax  kpsd  mdlb  ARO:3001084  tssj  icmftssm  acrr  \n",
      "0     0     0     0          0.0     0         0     0  \n",
      "1     0     0     0          0.0     0         0     0  \n",
      "2     0     0     0          0.0     0         0     0  \n",
      "3     0     0     0          0.0     0         0     0  \n",
      "4     0     0     0          0.0     0         0     0  \n",
      "\n",
      "[5 rows x 509 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Enterobacter hormaechei\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Escherichia coli\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Klebsiella pneumoniae\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Enterobacter hormaechei\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Escherichia coli\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Klebsiella pneumoniae\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Enterobacter hormaechei\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Escherichia coli\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Klebsiella pneumoniae\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Enterobacter hormaechei\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Escherichia coli\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Klebsiella pneumoniae\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Enterobacter hormaechei\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Escherichia coli\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Klebsiella pneumoniae\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Enterobacter hormaechei\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Escherichia coli\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
      "/tmp/ipykernel_2147886/4245227815.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
      "/tmp/ipykernel_2147886/4245227815.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"Species\"][i]= \"Klebsiella pneumoniae\"\n",
      "/tmp/ipykernel_2147886/4245227815.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n"
     ]
    }
   ],
   "source": [
    "#we use meta data information to add in our table\n",
    "concatenated_df2 = pd.concat([mysterious_df, merged_ETBCMet_df], axis=0, ignore_index=True).fillna(0)\n",
    "\n",
    "for i in range(len(concatenated_df2[\"ID\"]))[-18:]:\n",
    "    name = concatenated_df2[\"ID\"][i]\n",
    "    concatenated_df2[\"City\"][i]=name.split(\"_\")[0]\n",
    "    if name.split(\"_\")[1]==\"En\":\n",
    "        concatenated_df2[\"Species\"][i]= \"Enterobacter hormaechei\"\n",
    "        #\n",
    "        #eso si quieres quitar el nombre del genero de la ciudad\n",
    "    elif name.split(\"_\")[1]==\"Es\":\n",
    "        concatenated_df2[\"Species\"][i]= \"Escherichia coli\"\n",
    "        #concatenated_df2[\"ID\"][i]=name.split(\"_\")[0]\n",
    "    elif name.split(\"_\")[1]==\"Kl\":\n",
    "        concatenated_df2[\"Species\"][i]= \"Klebsiella pneumoniae\"\n",
    "        #concatenated_df2[\"ID\"][i]=name.split(\"_\")[0]\n",
    "    concatenated_df2[\"ID\"][i]=\"S\"+str(i+1)\n",
    "print(concatenated_df2.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2f17c014",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(concatenated_df2.tail(20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22b748e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
